### PR TITLE
REST API Fixes and Enhancement

### LINKED ISSUE #1373: BB REST API - Activity Endpoint : Field comment_count does not include sub comment (reply to comment)
**Describe the bug**
The field in the result of the rest api call below does not count the total number of comments. It only count the number of comments at the "first level" (not sub-comment).

`GET {{host}}/wp-json/buddyboss/v1/activity/813`

**To Reproduce**
1. Create an activity
2. Create a comment on that activity
3. Reply to the comment newly created (at step 2).
4. Make a rest api call to /wp-json/buddyboss/v1/activity/<id>
5. See the field "comment_count" (currently is 1 but should be 2).

**Expected behavior**
Add field (to the response) to get the count of all comments of an activity (# comments at level 1 + # comments at level 2 + # comments at level 3 + ...)

**Screenshots**
none

**Support ticket links**
none

### LINKED ISSUE #1424: BB REST API - Media Endpoint : the field activity_id is set to 0 when update a media
**Describe the bug**
When you want to "move" a media into an album, you need to make a Media Update Request. But, when you make that request, the field bp_media.activity_id is always reset to 0 (in the database).

`POST {{host}}/wp-json/buddyboss/v1/media/269?group_id=111&album_id=10`

This is because we can't set the parameter "activity_id" in an update request.

An easy solution could be to edit the line 432 of the file "bp-media-function.php"

From
`$media->activity_id   = (int) $r['activity_id'];`
To
`$media->activity_id   = ( ! $r['activity_id'] ? $media->activity_id : $r['activity_id']);`

**To Reproduce**
1. Take any media in the database with activity_id <> 0 (in the database bp_media)
2. Update that media (update album_id for example) by using a rest request.
3. Query the database to read bp_media.activity_id of your updated media.

**Expected behavior**
Do not reset activity_id when updating a media.

**Screenshots**
none

**Support ticket links**
none

### LINKED ISSUE #1741: BB REST API - Media Endpoint : Member can't list medias of his group because of permissions check issue
**Describe the bug**
There's an issue with the method _**get_items_permissions_check**_ in the file _**class-bp-rest-media-endpoint.php**_ (line 285). 

The condition `! in_array( $group->id, $user_groups['groups'], true )` returns true even when the user is a member because `$user_groups['groups']` contains strings and `$group->id` is an integer.

The condition at the line 285 should be : 

`} elseif ( ! bp_current_user_can( 'bp_moderate' ) && ! empty( $group->id ) && 'public' !== bp_get_group_status( $group ) && isset( $user_groups['groups'] ) && ! in_array( $group->id, $user_groups['groups'], false ) ) {`

Use strict = false in the php function `in_array`.

**To Reproduce**
1. Create a Private group
2. Make the rest request : `GET {{host}}/wp-json/buddyboss/v1/media?context=embed&page=1&per_page=42&group_id=152`

You should see : 

```
{
  "code": "bp_rest_authorization_required",
  "message": "Sorry, Restrict access to only group members.",
  "data": {
    "status": 403
  }
}
```
**Expected behavior**
Because user is a member of the group, he should be able to request medias of that group.

**Screenshots**
none

**Support ticket links**
none

### PR DESCRIPTION
**Pull request by KartikSuthar**
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

Fixes #1373
Fixes #1741 
Fixes #1424

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### CODE REVIEW

**Review by ExoGeek — COMMENTED** [review]
@KartikSuthar I've seen in the Projects dashboard that you plan to add the **_share_** functionnality. I don't know how you plan to implement that but the code from **line 603 to 628** of the function **create_item_permission_check** in the file **class-bp-rest-media-endpoint.php** may be a problem.

May I suggest to create a permission check as follow (for having access to a media) : 

1. The logged in user is the author of the attachment_id
2. **OR** the logged in user has already (read) access to the media (see bp_media.privacy) => this enable media sharing.

image

**Comment by KartikSuthar** [comment]
@ExoGeek Regarding your suggestions, 

1. I have already implemented `create_item_permission_check()` to check the current user is the author of the uploaded media here: https://github.com/buddyboss/buddyboss-platform/pull/1748/files#diff-ea9fff6b2037b919ceccab7dc7d1b711e39a1dd3f4e112b9d800d50daca3e264R603
2. We will update the API with the sharing module.

File: src/bp-document/classes/class-bp-rest-document-endpoint.php [hunk 1]
```diff
@@ -2461,4 +2510,24 @@ protected function bbp_document_update_rest_field_callback( $object, $value ) {
...
+	 */
+	public function bp_rest_attachment_document_id( $attachment_id = 0 ) {
+		global $bp, $wpdb;
+
+		if ( ! $attachment_id ) {
+			return false;
+		}
+
+		// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.DirectDatabaseQuery
+		$attachment_document_id = (int) $wpdb->get_var( "SELECT DISTINCT d.id FROM {$bp->document->table_name} d WHERE d.attachment_id = {$attachment_id}" );
```
Comment (chetansatasiya) [inline src/bp-document/classes/class-bp-rest-document-endpoint.php:2528]: @KartikSuthar Create a seprate function so if anyone need document id via the attachment id they can get and not makea seprate query everywhere.

**Review by chetansatasiya — CHANGES_REQUESTED** [review]
@KartikSuthar Some places still spaces is not correct that need to make a proper.

File: src/bp-media/classes/class-bp-rest-media-albums-endpoint.php [hunk 2]
```diff
@@ -905,6 +905,8 @@ public function prepare_item_for_response( $album, $request ) {
 			'title'         => $album->title,
 			'privacy'       => $album->privacy,
 			'media'         => $album->media,
+			'group_name'    => ( isset( $album->group_name ) ? $album->group_name : '' ),
```
Comment (ExoGeek) [inline src/bp-media/classes/class-bp-rest-media-albums-endpoint.php:908]: Do not forget to add fields **visibility** and **group_name** into the method **get_item_schema**.

File: src/bp-media/classes/class-bp-rest-media-endpoint.php [hunk 3]
```diff
@@ -1260,6 +1310,8 @@ public function prepare_item_for_response( $media, $request ) {
 			'menu_order'      => $media->menu_order,
 			'date_created'    => $media->date_created,
 			'attachment_data' => $media->attachment_data,
+			'group_name'      => ( isset( $media->group_name ) ? $media->group_name : '' ),
```
Comment (ExoGeek) [inline src/bp-media/classes/class-bp-rest-media-endpoint.php:1313]: Do not forget to add fields **visibility** and **group_name** into the method **get_item_schema**.

File: src/bp-media/classes/class-bp-rest-media-albums-endpoint.php [hunk 4]
```diff
@@ -905,6 +905,8 @@ public function prepare_item_for_response( $album, $request ) {
...
+			'visibility'    => ( isset( $album->visibility ) ? $album->visibility : '' ),
 			'user_email'    => $album->user_email,
```
Comment (ExoGeek) [inline src/bp-media/classes/class-bp-rest-media-albums-endpoint.php:910]: Just a question : why this field "user_email" is added in the response. The email can be private and shouldn't be exposed so easily...

**Review by ExoGeek — APPROVED** [review]
It works ! Thanks.

File: src/bp-activity/bp-activity-filters.php [hunk 5]
```diff
@@ -1721,7 +1721,7 @@ function bp_activity_media_add( $media ) {
 function bp_activity_create_parent_media_activity( $media_ids ) {
 	global $bp_media_upload_count, $bp_activity_post_update, $bp_media_upload_activity_content, $bp_activity_post_update_id, $bp_activity_edit;
 
-	if ( ! empty( $media_ids ) && empty( $bp_activity_post_update ) && ! isset( $_POST['edit'] ) ) {
+	if ( ! empty( $media_ids ) && empty( $bp_activity_post_update ) && ! isset( $_POST['edit'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Missing
```
Comment (chetansatasiya) [inline src/bp-activity/bp-activity-filters.php:1724]: @KartikSuthar use the `filter_input` to sanitize the `edit`

File: src/bp-activity/bp-activity-filters.php [hunk 6]
```diff
@@ -2034,7 +2042,7 @@ function bp_activity_document_add( $document ) {
 function bp_activity_create_parent_document_activity( $document_ids ) {
 	global $bp_document_upload_count, $bp_activity_post_update, $bp_document_upload_activity_content, $bp_activity_post_update_id, $bp_activity_edit;
 
-	if ( ! empty( $document_ids ) && empty( $bp_activity_post_update ) && ! isset( $_POST['edit'] ) ) {
+	if ( ! empty( $document_ids ) && empty( $bp_activity_post_update ) && ! isset( $_POST['edit'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Missing
```
Comment (chetansatasiya) [inline src/bp-activity/bp-activity-filters.php:2045]: @KartikSuthar Use sanitize on `edit`

**Review by chetansatasiya — APPROVED** [review]
@KartikSuthar 2 places needs enahancement. other then Looks good to me..!!

**Review by mehulkaklotar — APPROVED** [review]

**Pull request merged**